sympy/solvers/solveset.py

```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
     # Return solutions
     solution = FiniteSet(tuple(solution))
     return solution
+
+
+##############################################################################
+# ------------------------------nonlinsolve ---------------------------------#
+##############################################################################
+
+
+def substitution(system, symbols, result, known_symbols, exclude, all_symbols):
+    r"""
+    Solves the `system` using substitution method.
+
+    Parameters
+    ==========
+
+    system : list of equations.
+
+    symbols : list of unsolved symbols.
+
+    known_symbols : list of solved symbols.
+
+    result : If non of the symbols are solved then it is empty list, otherwise
+    list of already solved_syms symbols( dict symbol : value )
+
+    exclude : Set of expression which should not be zero with result.
+    Mostly denominator expression.
+
+    all_symbols : known_symbols + unsolved symbols.
+
+    Returns
+    =======
+    Returns a FiniteSet of ordered tuple.
+
+    Examples
+    ========
+
+    >>> from sympy.core.symbol import symbols
+    >>> x, y = symbols('x, y', real = True)
+    >>> from sympy.solvers.solveset import substitution
+    >>> substitution([x +y], [x], [{y : 1}], [y], set([]), [x, y])
+    {(-1, 1)}
+    >>> substitution([x +y], [x], [{y : 1}], [y], set([x + 1]), [y, x])
+    EmptySet()
+    >>> substitution([x +y], [x], [{y : 1}], [y], set([x - 1]), [y, x])
+    {(1, -1)}
+
+    """
+
+    from sympy.core.compatibility import ordered, default_sort_key
+    from sympy import Complement
+    from sympy.core.containers import Dict
+
+    def _ok_syms(e, sort=False):
+        # get the unsolved symbol.
+            rv = (e.free_symbols - set(known_symbols)) & set(all_symbols)
+            if sort:
+                rv = list(rv)
+                rv.sort(key=default_sort_key)
+            return rv
+
+    # sort such that equation with the fewest potential symbols is first
+    for eq in ordered(system, lambda _: len(_ok_syms(_))):
+        newresult = []
+        got_s = set()
+        hit = False
+        complements = {}
+        for r in result:
+            # update eq with everything that is known so far
+            eq2 = eq.subs(r)
+            ok_syms = _ok_syms(eq2, sort=True)
+            if not ok_syms:
+                if r:
+                    newresult.append(r)
+                break  # skip as it's independent of desired symbols
+            for s in ok_syms:
+                not_solvable = False
+                soln_imageset = None
+                try:
+                    soln = solveset_real(eq2, s)
+                    # If we want to add the complex solution also then
+                    # do the same with soln_complex.
+                    # soln_complex = solveset_complex(eq2, s)
+                    if not soln:
+                        soln = solveset_complex(eq2, s)
```

Comment:
What's the motivation behind solving in complex domain only when there is no solution in real domain?